core/genesis/file.go

```diff
@@ -0,0 +1,251 @@
+package genesis
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/protobuf/proto"
+)
+
+
+type FileGenesisSource struct {
+	sourceDir  string
+	moduleName string
+}
+
+const (
+	fileOpenflag  = os.O_CREATE | os.O_WRONLY
+	flieOpenMode  = fs.FileMode(0o600)
+	dirCreateMode = fs.FileMode(0o700)
+)
+
+// NewFileGenesisSource returns a new GenesisSource for the provided
+// source directory and the provided module name where it is assumed
+// that the encoded json data of file.
+func NewFileGenesisSource(sourceDir, moduleName string) GenesisSource {
+	return &FileGenesisSource{sourceDir: filepath.Clean(sourceDir), moduleName: moduleName}
+}
+
+// OpenReader opens the source file reading from the given parameters,
+// and returns a ReadCloser.
+// It will try to open the file in order following by:
+// <sourceDir>/<module>/<field>.json
+// <sourceDir>/<module>.json
+// <sourceDir>/genesis.json
+func (f *FileGenesisSource) OpenReader(field string) (io.ReadCloser, error) {
+	// try reading genesis data from <sourceDir>/<module>/<field>.json
+	fName := fmt.Sprintf("%s.json", field)
+	fPath := filepath.Join(f.sourceDir, f.moduleName)
+
+	fp, err := os.Open(filepath.Clean(filepath.Join(fPath, fName)))
+	if err == nil {
+		return fp, nil
+	}
+
+	// if cannot find it, try reading from <sourceDir>/<module>.json
```

Comment:
Once we get to this point we should just call `f.ReadRawJSON` and just wrap the result in a `bytes.Reader`